library/ecdh.c

```diff
@@ -91,64 +102,161 @@ int mbedtls_ecdh_compute_shared( mbedtls_ecp_group *grp, mbedtls_mpi *z,
  */
 void mbedtls_ecdh_init( mbedtls_ecdh_context *ctx )
 {
+#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
     memset( ctx, 0, sizeof( mbedtls_ecdh_context ) );
+
+    mbedtls_ecp_group_init( &ctx->grp );
+    mbedtls_ecp_point_init( &ctx->Q   );
+    mbedtls_ecp_point_init( &ctx->Qp  );
+#else
+    memset( ctx, 0, sizeof( mbedtls_ecdh_context ) );
+#endif
+}
+
+static int mbedtls_ecdh_setup_internal( mbedtls_ecdh_context_mbed *ctx,
+                                        mbedtls_ecp_group_id grp )
+{
+    int ret;
+
+    ret = mbedtls_ecp_group_load( &ctx->grp, grp );
+    if( ret != 0 )
+    {
+        return( MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE );
+    }
+
+    return( 0 );
 }
 
 /*
- * Free context
+ * Setup context
  */
-void mbedtls_ecdh_free( mbedtls_ecdh_context *ctx )
+int mbedtls_ecdh_setup( mbedtls_ecdh_context *ctx, mbedtls_ecp_group_id grp )
 {
     if( ctx == NULL )
-        return;
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
+    return( mbedtls_ecdh_setup_internal( ctx, grp ) );
+#else
+    switch( grp )
+    {
+        default:
+            ctx->point_format = MBEDTLS_ECP_PF_UNCOMPRESSED;
+            ctx->var = MBEDTLS_ECDH_VARIANT_MBEDTLS_2_0;
+            ctx->grp = grp;
+            return( mbedtls_ecdh_setup_internal( &ctx->ctx.mbed_ecdh, grp ) );
+    }
+#endif
+}
 
+static void mbedtls_ecdh_free_internal( mbedtls_ecdh_context_mbed *ctx )
+{
     mbedtls_ecp_group_free( &ctx->grp );
     mbedtls_ecp_point_free( &ctx->Q   );
     mbedtls_ecp_point_free( &ctx->Qp  );
-    mbedtls_ecp_point_free( &ctx->Vi  );
-    mbedtls_ecp_point_free( &ctx->Vf  );
     mbedtls_mpi_free( &ctx->d  );
     mbedtls_mpi_free( &ctx->z  );
-    mbedtls_mpi_free( &ctx->_d );
```

Comment:
The legacy context still contains `Vi`, `Vf`, and `_d`. Shouldn't we still free these for the legacy context configuration case? (Perhaps free these in `mbedtls_ecdh_free()` before the call to `mbedtls_ecdh_free_internal()` in the legacy configuration case.)